substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,980 @@
+
+	/// The configuration for the next epoch.
+	#[pallet::storage]
+	#[pallet::getter(fn next_config)]
+	pub type NextEpochConfig<T> = StorageValue<_, EpochConfiguration>;
+
+	/// Pending epoch configuration change that will be set as `NextEpochConfig` when the next
+	/// epoch is enacted.
+	///
+	/// In other words, a config change submitted during epoch N will be enacted on epoch N+2.
+	/// This is to maintain coherence for already submitted tickets for epoch N+1 that where
+	/// computed using configuration parameters stored for epoch N+1.
+	#[pallet::storage]
+	pub(super) type PendingEpochConfigChange<T> = StorageValue<_, EpochConfiguration>;
+
+	/// Stored tickets metadata.
+	#[pallet::storage]
+	pub type TicketsMeta<T> = StorageValue<_, TicketsMetadata, ValueQuery>;
+
+	/// Tickets identifiers.
+	///
+	/// The key is a tuple composed by:
+	/// - `u8` equal to epoch-index mod 2
+	/// - `u32` equal to the slot-index.
+	#[pallet::storage]
+	pub type TicketsIds<T> = StorageMap<_, Identity, (u8, u32), TicketId>;
+
+	/// Tickets to be used for current and next epoch.
+	#[pallet::storage]
+	pub type TicketsData<T> = StorageMap<_, Identity, TicketId, TicketBody>;
+
+	/// Next epoch tickets accumulator.
+	///
+	/// Contains lists of tickets where each list represents a burst of tickets received
+	/// via the `submit_tickets` extrinsic.
+	/// Given that each list max capacity is bounded by the max tickets we can handle
+	/// in one epoch, there's enough space for tickets here.
+	///
+	/// Special `u32::MAX` key is reserved for the most recently sorted segment.
+	#[pallet::storage]
+	pub type NextTicketsSegments<T: Config> =
+		StorageMap<_, Identity, u32, BoundedVec<TicketId, MaxTicketsFor<T>>, ValueQuery>;
+
+	/// Parameters used to verify tickets validity via ring-proof.
+	///
+	/// In practice: Updatable Universal Reference String and the seed.
+	#[pallet::storage]
+	#[pallet::getter(fn ring_context)]
+	pub type RingContext<T: Config> = StorageValue<_, vrf::RingContext>;
+
+	/// Genesis configuration for Sassafras protocol.
+	#[pallet::genesis_config]
+	#[derive(frame_support::DefaultNoBound)]
+	pub struct GenesisConfig<T: Config> {
+		/// Genesis authorities.
+		pub authorities: Vec<AuthorityId>,
+		/// Genesis epoch configuration.
+		pub epoch_config: EpochConfiguration,
+		/// Phantom config
+		#[serde(skip)]
+		pub _phantom: sp_std::marker::PhantomData<T>,
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+		fn build(&self) {
+			Pallet::<T>::initialize_genesis_authorities(&self.authorities);
+			EpochConfig::<T>::put(self.epoch_config.clone());
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// Block initialization
+		fn on_initialize(now: BlockNumberFor<T>) -> Weight {
+			// Since `initialize` can be called twice (e.g. if an external session manager is used)
+			// let's ensure that we only do the initialization once per block.
```

Comment:
Wow, I didn't know this! Is it documented somewhere? `on_initialize` doesn't mention that it can be called more than once, I'd be very surprised if that happened.

Reply:
That can happen when new epoch enactment is driven by an external pallet (e.g. session-pallet).

As this PR was not supposed to support it, I removed that code snip.

BTW I can still reference BABE for such behavior. 

The external component can query the pallet to know if he can end the session (calls into  [this](https://github.com/paritytech/polkadot-sdk/blob/841a33ede4c991521d5c3ebf852170e978660742/substrate/frame/babe/src/lib.rs#L516-L522)).

Pallet session invokes the `should_end_session` in its `on_initialize` hook ([here](https://github.com/paritytech/polkadot-sdk/blob/841a33ede4c991521d5c3ebf852170e978660742/substrate/frame/session/src/lib.rs#L564)).

To give a response the pallet needs to initialize its stuff, and thus calls `initialize` (not using the `on_initialize` hook).

Because is not guaranteed that which `on_initialize` hook will be called first (pallet-session or pallet-babe) and the `should_end_session` may be called first then in the babe `on_initialize` hook we need a way to check if it has already been called.

I know that may be a bit hackish, but is the current solution.

Reply:
As for your note about the documentation. I'd say that this is a completely internal machinery of the implementation.

Is not the `on_initialize` hook that is called twice. Is babe that allows an external component to call its `initialize` function and (knowing this possibility) it internally prevents a double call to the `initialize` function.